Mathlib/GameTheory/Auction/Basic.lean

```diff
@@ -0,0 +1,193 @@
+
+## Main Results
+
+- `utility_nneg`: Utility is non-negative if the bid equals the valuation in second price auction.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy in second price auction.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in first price auction.
+
+## Helping Lemmas
+
+- `eq_winner_of_bid_gt`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_maxBid`: There exists a participant whose bid matches the highest bid
+- `bid_winner_eq_maxBid`: The winner's bid is the highest.
+- `winnerbid_ge_second`: The winner's bid is at least the second highest bid.
+- `Secondprice.utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `Secondprice.utility_loser`: If `i` does not win, their utility is 0.
+- `bid_le_bid_winner`: The winner's bid is greater than or equal to all other bids.
+- `maxBidExcluding_eq_maxBid_if_loser`: If `i` does not win, the highest bid excluding `i` equals
+   the highest bid.
+- `Firstprice.utility_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `Firstprice.utility_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxBid` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+/- `v` is the valuation for participants `I` -/
+variable {I : Type*} [Fintype I] [Nontrivial I] (v : I → ℝ)
+
+namespace Auction
+
+/- `b` is the bidding function that maps participant to a real number bid. -/
+variable (b : I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxBid : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
+
+/-- There exists a participant `i` whose bid equals the highest bid. -/
+lemma exists_maxBid : ∃ i : I, b i = maxBid b := by
+   obtain ⟨i, _, h2⟩ := Finset.exists_mem_eq_sup' Finset.univ_nonempty b
+   exact ⟨i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : I := Classical.choose (exists_maxBid b)
+
+/-- The bid of the winner equals the highest bid. -/
+lemma bid_winner_eq_maxBid : b (winner b) = maxBid b := Classical.choose_spec (exists_maxBid b)
+
+/-- The bid of the winner is always greater than or equal to the bids of all others.-/
+lemma bid_le_bid_winner (j : I) : b j ≤ b (winner b) := by
+  rw [bid_winner_eq_maxBid b]
+  exact Finset.le_sup' b (Finset.mem_univ j)
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma eq_winner_of_bid_gt (i : I) (H : ∀ j , j ≠ i → b j < b i) : i = winner b := by
```

Comment:
`lt` might be a better name since that is what appears in the statement of the lemma?
```suggestion
lemma eq_winner_of_bid_lt (i : I) (H : ∀ j , j ≠ i → b j < b i) : i = winner b := by
```